File: android/app/src/main/java/it/airgap/wallet/plugin/isolatedmodules/js/environment/JSEnvironment.kt
package it.airgap.wallet.plugin.isolatedmodules.js.environment

import com.getcapacitor.JSObject
import it.airgap.wallet.plugin.isolatedmodules.js.JSModule
import it.airgap.wallet.plugin.isolatedmodules.js.JSModuleAction
import it.airgap.wallet.util.JSException
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

interface JSEnvironment {
    suspend fun isSupported(): Boolean
    @Throws(JSException::class)
    suspend fun run(module: JSModule, action: JSModuleAction, ref: String? = null): JSObject
    suspend fun reset(runRef: String)
    suspend fun destroy()

    abstract class SandboxRegistry<T> {
        val mutex: Mutex = Mutex()
        val sandboxes: MutableMap<String, T> = mutableMapOf()

        suspend inline fun getOrPut(module: JSModule, defaultValue: () -> T): T =
            mutex.withLock {
                sandboxes.getOrPut(module.identifier, defaultValue)
            }

        suspend fun reset() {
            mutex.withLock {
                sandboxes.apply {
                    values.forEach { it.reset() }
                    clear()
                }
            }
        }

        protected abstract suspend fun T.reset()
    }

    enum class Type {
        WebView, JavaScriptEngine;

        companion object {
            fun fromString(value: String): Type? = values().find { it.name.lowercase() == value.lowercase() }
        }
    }
}

File: ios/App/App/IsolatedModules/JS/Environment/JSEnvironment.swift
//
//  JSEnvironment.swift
//  App
//
//  Created by Julia Samol on 07.02.23.
//

import Foundation

protocol JSEnvironment {
    func run(_ action: JSModuleAction, in module: JSModule, ref: String?) async throws -> [String: Any]
    func reset(runRef: String) async throws
    func destroy() async throws
}

enum JSEnvironmentKind: String, Codable {
    case webview
}

extension JSEnvironment {
    func run(_ action: JSModuleAction, in module: JSModule) async throws -> [String: Any] {
        try await run(action, in: module, ref: nil)
    }
}
